docs: klip-9: Table Functions

purplefox opened the pull request:
### Description 

Fixes: https://github.com/confluentinc/ksql/issues/3512

This PR contains a new KLIP - KLIP-9 proposal for table functions.

### Testing done 

Non functional change

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

MichaelDrogalis commented on design-proposals/klip-9-table-functions.md:
```diff
@@ -0,0 +1,206 @@
+We would like to implement
+
+* Table functions on the SELECT clause only
+* Multiple table functions on the SELECT clause
```
What would be the semantics for 2+ table functions? Would it work like a zipmap function (e.g. zip a new row per index in each collection)?

purplefox replied:
Yep, exactly like a zipmap. There's a bit of detail about this in the design section.

MichaelDrogalis replied:
Ah, there it is. I foolishly skipped the design section while reading. :)

apurvam replied:
The other option here is to return a cross product when there are multiple table valued functions. 

So if you had functions `func1(col1)` and `func2(col2)` returning rows `[x, y]` and `[a, b]` respectively, you could output:

```
x, a
x, b
y, a
y, b
```

The current proposal would return

```
x, a
y, b
```

Did we consider supporting both semantics? Is there any value in doing so?

purplefox replied:
At least Hive and PostgreSQL support UDTFs on SELECT clause, and this seems the simplest way to express a zipjoin across multiple UDTFs on the select. A zipjoin seems the most useful use case.
Expressing a zipjoin in SQL on a JOIN is very non-unintuitive IMHO (it requires correlated subqueries). I think there's a comment and illustration to that effect on the original discussion.
I don't the cartesian product is terribly useful compared to a zipjoin.

derekjn commented on design-proposals/klip-9-table-functions.md:
```diff
@@ -0,0 +1,206 @@
+of type `V` holding the value of the element
+* `EXPLODE(MAP<K, V>)` -> Results in `N` rows (where `N` is number of elements in map) each with a column
+holding a `STRUCT<key K, value V>` that contains the key and value.
+* `INDEX_EXPLODE(ARRAY<V>)` -> Results in `N` rows (where `N` is number of elements in array) each with a
```
This is a cool idea, although I'm wondering if we'd be pulling too much functionality into `explode` with this (I understand it's a separate function, but nevertheless). That is, could there be a way to achieve this result using an intuitive composition of simpler functions instead? For example, we could have a `series` (table!) function that could be used to enumerate an array like so:

```sql
-- Example event:
-- {"arr": [0, 1, 2]}

SELECT series(0, len(arr)) AS index, explode(arr) FROM stream EMIT CHANGES;
index | explode
----------------
  0   |   0
  1   |   1
  2   |   2
```
Or even something that wraps the above array length check:
```sql
SELECT enumerate(arr), explode(arr) ...
```
In general I believe we'll be able to provide users with greater flexibility by solving problems like these using compositions of simple functions. Good composability is also conducive to solving unforeseen problems/use cases that we didn't necessarily design for.

purplefox replied:
This particular function is borrowed from Hive (pos_explode). I agree with the point about composability but we already have an explode method which doesn't provided indexes. And I think functions like series() would be a good idea, so they could just be used together to achieve the same ends.
I think sometimes though it's useful to provide some more denormalised functions for commonly used, use cases. I'm guessing that getting the index is going to be one of those, but it's just a guess! ;) It's the same kind of reasoning of why HashMap in Java has an entries() collection as well keys() and values() ones. It's not strictly necessary but it's so often the case you want to iterate the key and value together that it can lead to clearer code. Having said that I'm not hung up on this function and will remove it if people don't think it's useful.

big-andy-coates replied:
IMHO I see the benefit of pos_explode, but if given the choice between `series` _or_ `pos_explode` I'd take the former any day.  

So maybe we should consider either replacing `pos_explode` in this doc with `series` or adding `series` as well, but not having `pos_explode` without `series`.

derekjn commented on design-proposals/klip-9-table-functions.md:
```diff
@@ -0,0 +1,206 @@
+Table functions are typically used in FROM clauses with traditional databases. Some databases
+such as PostgreSQL and Hive also allow table functions to appear in SELECT clauses.
+
+An example of a table function from Apache Hive is `explode` - this takes a non scalar type such as
```
Are we decided on using `explode` as the name? As far as I know, `explode` arose from looking at Hive as an example and the name just kind of stuck naturally from there. Personally I don't have super strong feelings about `explode` versus `unnest`, which is probably the other most viable option. I find `unnest` more descriptive but I'm fine with either option.

We don't need to spend time on naming this since both options are pretty good, I just thought I'd call out the fact that we do have choices here.

purplefox replied:
Naming is hard! ;)
Personally I find explode more descriptive and I think it evokes better what the function does.
unnesst to me is a bit too much of a software engineering (or maths) term. (and it's not clear to me why unnest means going from one value to zero or more rows) If our target demographic of users is not so technical folks, then my preference would be explode.
Again though, I don't have very strong feelings and we can change easily before release anyway :)

purplefox replied:
Another choice of name which I think is intuitive is `flatten`

derekjn replied:
For me `explode` evokes the idea of expanding one (scalar) value into many values, which isn't exactly what this is doing. I think `unnest` comes from the idea of nested rows, as an array can be considered a row within a row.

hpgrahsl replied:
I know I'm a little bit late with commenting here... just adding my 2 cents by comparing how others have approached this:
* `explode` is also used, e.g. in SparkSQL. BTW they also differentiate between `explode` (just creating multiple rows) vs. `posexplode` (...with indexes/position columns) (see https://spark.apache.org/docs/2.4.0/api/sql/index.html#explode
* interestingly folks at MS have chosen a different approach with ASA and offer CROSS and OUTER APPLY functionality in this regard (see https://docs.microsoft.com/en-us/stream-analytics-query/apply-azure-stream-analytics)

purplefox commented:
Responding to @apurvam 's point about cross product.

I did consider providing cross product semantics at the beginning, but after conversations with @derekjn and others it became clearer to me that a "zip" join (is there a better term for this? ;) ) would probably be more useful.

E.g. I think data in the following kind of form

```
{"country":"usa",customer_names:["tim", "derek", "apurva"], customer_ages[99, 98, 97]}
```

exploding to:
```
{"country":"usa", customer_name:"tim", customer_age:99}
{"country":"usa", customer_name:"derek", customer_age:98}
{"country":"usa", customer_name:"apurva", customer_age:97}

```

is going to be more common than doing a cross join.

apurvam commented:
Thanks for sharing the rationale @purplefox. That seems reasonable. If we find that cross join semantics are desirable in the future, what are are options for introducing those semantics incrementally? 

It would be good to think through that, because I imagine we will find valid use cases for cross joins.

purplefox commented:
@apurvam One way of introducing cross join semantics for table functions would be to create a table function called CROSS_JOIN which took two ARRAYs as parameters and did the cross join in the function, returning the cross joined results. This would be doable with the current proposal without any changes.
E.g.
`SELECT CROSS_JOIN(EXPLODE(f1), EXPLODE(f2)) from S1`
Or we could call the function `MULTIPLY_ARRAY` (might be better)

hjafarpour commented:
@purplefox I'm also in favor of the cross join as the default behavior instead of the current one. You can achieve the current behavior by using `INDEX_EXPLODE` and filtering the result of the cross join case based on the indexes easily.
Also, in the current behavior you assume the input collections have a certain order, although this is correct for ARRAY inputs but it does not apply for MAP inputs so you would have non deterministic results if you have MAPs and use two EXPLODE functions.

derekjn commented:
I'm not sure that cross joining would be semantically correct for multiple `explode` calls in the `SELECT` expression list. Cross join behavior seems like it would be more appropriate for multiple `explode` calls in a `FROM` list, IMO.

purplefox commented:
I think zipjoin semantics for multiple table functions in a select seems more useful and most likely what most users would want/expect, so I think we should be pragmatic and optimise for what most users are likely to want.
As yet, I haven't heard a convincing use case that would need cross join semantics, however if there is demand then it should be fairly easily supported by writing a new table function which does the cross product of the two arrays.

hjafarpour commented:
The question is what would be the result of having two explode calls with `MAP` inputs. How would you determine the order? 
On writing a table function to have the cross join output, note that the result of the function will be one column while the result of the cross join for two explodes will have two columns.

agavra commented on design-proposals/klip-9-table-functions.md:
```diff
@@ -0,0 +1,206 @@
+
+There will be no changes to the KSQL language or any existing public APIs or configuration files.
+
+We will introduce new annotations to allow users to mark their UDTFs, in a very similar way to
```
it'd be nice to spell this out a little bit as there might be interesting consequences based on how we choose to model it. e.g. if it returns an `Iterable<TYPE>` vs a `List<TYPE>` there may be memory/performance implications we need to consider (e.g. the Iterable approach would open the door to lazy computation)

There's also a consideration about infinite returns. Do we want to support that? e.g. if we have an iterable instead of a list or an array, then a user could implement something like `RANDOMS()` which will infinitely output rows for one input.

purplefox replied:
Hey @agavra Thank you for raising such an interesting point!

Sorry in advance for the long post, but this is one of my specialist areas.

In an ideal world I would love us to be able to provide functions that produce infinite results, or produce results asynchronously or lazily-  this opens us up to a lot more interesting use cases that could be implemented.

However for this to work it needs support from the underlying streams library. AIUI, right now the Kafka Streams DSL (fFlatMap) uses a synchronous blocking approach (i.e. “please give me the entire result of the FlatMap and I will block waiting for it).

See https://github.com/apache/kafka/blob/trunk/streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamFlatMapValues.java#L40

I don't think that's going to work for slow, infinite or async functions. It also doesn’t seem obvious how we could write our own custom Kafka Streams processor to do this.

To make a really scalable event processing platform based on the pull model (polling) you need some way of ensuring that calls to getNext() (e.g. on an iterator) never block. Blocking kills performance as the blocking thread can’t do other work while it’s blocked. For a timed iterator that only creates, say, one new result every 10 seconds, this would be a disaster. When there are many running queries in the server blocking, you’d end up with large number of threads just to keep the system responsive, plus RAM overhead per thread, not to mention overhead from extra context switching.

To avoid blocking on a poll(), scalable event processing systems (Rust futures are a good example of this) need some way of asking the task whether it’s ready or not so a poll doesn’t block, e.g. an isReady() method. If isReady() returns false, then the task needs to be parked, and the task also needs to wake itself up sometime in the future and reschedule itself when it has more data to provide. To cut a long story short, it gets complicated! A simple iterator is not going to help us here, the API of the function would need to support the non blocking semantics and it also needs these concepts built into the underlying stream processing framework. AIUI the blocking architecture of Kafka Streams is really not compatible with that way of working.

An alternative way of implementing this scalable is using an async model (push model). In this model there is no polling happening to see if results are ready. Instead results are provided asynchronously to the caller (usually via a callback function or a future) when they are ready. In this model there is no blocking either so it can scale well, but it also requires support from the underlying streams framework to work. With the push model, it also needs to handle back pressure (aka flow control), e.g. to prevent very fast producers of messages overwhelming things downstream and causing OOMs. This model is used quite widely including Vert.x, ReactiveX, Spring Reactor etc.

To summarise: I would love slow/infinite/async functions, but I don’t see any way of implementing these efficiently/scalably using Kafka Streams using the DSL or Processor APIs (perhaps there is another API I am not aware of?) And even if we’re not tied to Kafka Streams the choice of API for the function would depend on whether the underlying streams framework uses a push or pull model.

big-andy-coates replied:
I may be wrong, as I'm not KS expert, but I think having a infinite UDTF would be possible with the Processor API of KS, as it does not expect all the results in one go, (unlike `flatmap`), instead you can call `ProcessorContext#forward(k, v)` as often as you like.

However, I'm not sure how gracefully KS would handle such a case. I have a feeling any infinite UDTF would cause the KS consumers to be ejected from the consumer group as they can't poll Kafka until the UDTF returns.

cc @guozhangwang @mjsax  @bbejeck  to confirm.

That said, does it actually make much sense to use an infinite UDTF on a table/stream, i.e. topic, source? Essentially, only the first record per-partition would be read, and then the infinite output would commence. As consumers were ejected from their groups those partitions may be assigned to stream threads which previously had no tasks, (threads with tasks will be stuck output infinite rows, so won't have new tasks assigned). As the partition moved to another stream thread it too would read the first message and commence infinite output.  So, the result would be inconsistent as we could have more than one stream thread performing infinite output per source partition.

That is not to say that infinite UDTFs are not of use, only that they don't make sense on a infinite source.  Maybe they do make sense on a finite source, e.g. using Oracle's `DUAL` source equivalent:

```sql
SELECT myInfiniteUdtf("blah') FROM DUAL;
```

Or using another UDTF to generate a finite source:

```sql
-- kick off datagen for the three types provided.
SELECT myDataGen(type) FROM explode("users", "clicks", "status-codes") AS type;
```

I'm not suggesting we support all of this / any of this now, though we can think about if we want to. Regardless, we should at least consider if our API can be designed, as @agavra says, to support such a thing in the future. 

As for the comments about async/sync push/pull threading etc.  These are obviously all valid points, but separate to the design of our API for UDTFs.  Having a thread per infinite/slow/async UDTF may be inefficient, but we can always later extend the framework from, say, UDTFs that return `Iterator<String>` to `Iterator<Future<String>>` to allow us to handle such requests more efficiently.

agavra replied:
@big-andy-coates I brought up infinite UDTFs more as a way of demonstrating how the API can be used (or maybe abused) if we designed it in one way vs another, so I think the discussion is important when we are designing our APIs. If we decide that infinite (or more specifically, anything that's non-immediate) UDTFs are valuable, then we need to support some iterable-type (as @purplefox pointed out, `Iterator` might not be smart) interface as opposed to one that's set in stone.

With regards to whether or not it's possible, that's also something I'm not familiar enough to comment on (waiting for the streams folk to chime in here), but I think if we decided that it's something that we want we should design our API around that, potentially updating KS to support it later.

purplefox replied:
> I may be wrong, as I'm not KS expert, but I think having a infinite UDTF would be possible with the Processor API of KS, as it does not expect all the results in one go, (unlike `flatmap`), instead you can call `ProcessorContext#forward(k, v)` as often as you like.

I am also not a KS expert but my understanding is there would be a thread blocking in KS until all the results from the infinite or slow UDTF are returned, and for infinite stream that never occurs. That doesn't seem very scalable.

> That is not to say that infinite UDTFs are not of use, only that they don't make sense on a infinite source. 

Right I don't think the suggestion was for them to work on an infinite source, more like things that can create a stream of random numbers of timing events (a tick every 10 seconds for example).

> As for the comments about async/sync push/pull threading etc. These are obviously all valid points, but separate to the design of our API for UDTFs. Having a thread per infinite/slow/async UDTF may be inefficient, but we can always later extend the framework from, say, UDTFs that return `Iterator<String>` to `Iterator<Future<String>>` to allow us to handle such requests more efficiently.

That API very much depends on what approach (push/pull) is taken. With a polling approach it would also be necessary to signal readiness, park and reschedule tasks - it's complex! For an async approach a simple future probably wouldn't suffice as we would need some way of propagating back pressure. All these considerations would need to be built into the API.

guozhangwang replied:
The current threading model of KS would indeed block the consumer from calling "poll" before a record has "completely" traversed through the topology, i.e. all its inferred records from the operator has been processed.

However, Streams team have some discussions about refactoring the threading model https://confluentinc.atlassian.net/wiki/spaces/KSTREAMS/pages/893749214/2019-06-07+and+more+Deep+Dive+on+Dynamic+Scaling, the main goal is to saturate the resources of a single node, but as a by-product it also means async-processing from the pov of the main thread calling poll, so infinite results it would not be a problem for blocking, BUT they would still be tricky from the committing aspect, as you still have to decide when it is safe to commit that "all records till this boundary has been processed".

purplefox replied:
Thanks for the insight @guozhangwang I will follow the discussion on the KS threading model changes :)

agavra commented on design-proposals/klip-9-table-functions.md:
```diff
@@ -0,0 +1,206 @@
+
+We will need the following general plumbing to support table functions:
+
+* In a similar way to how UDAFs are currently supported we will need to create a 
```
unless I'm missing something, UDTFs take scalar inputs right and don't need an undo/merge? In that case, can we model them after UDFs instead of after UDAFs? I think the model is easier to understand (there is one `@Udf` method for each UDF and no `Factory`).

agavra commented on design-proposals/klip-9-table-functions.md:
```diff
@@ -0,0 +1,206 @@
+There will be no changes to the KSQL language or any existing public APIs or configuration files.
+
+We will introduce new annotations to allow users to mark their UDTFs, in a very similar way to
+UDAFs.
```
should we have any concept of timing? e.g. if I have an input and it takes me a long time to compute each output, do I need to do it all at once or can I output one row at a time?

purplefox replied:
I could be wrong, but I can't see any way of processing partial results without blocking the caller using Kafka Streams. (See previous comment)

big-andy-coates replied:
As per previous comment ;) I think this may be possible if we use a KS `Processor`.

purplefox replied:
I'm guessing you can't call ProcessorContext.forward for a message after the original call to Processor.process() has returned. Otherwise the upstream processor which is actually reading from the consumer will think the message has been processed before its actually been seen by all downstream processors, and update offsets thus changing delivery semantics (it won't be at least once any more).

purplefox replied:
So you'd need to block processor.process, and that's not going to scale when you've got 1000 queries running in the server. Not to mention the issue you brought up about consumers getting thrown out of consumer groups.

purplefox commented:
Hi @hjafarpour 

> The question is what would be the result of having two explode calls with `MAP` inputs. How would you determine the order?

I don't think the lack of order with a Map has anything to do with table function joining semantics: EXPLODE on a single Map will not provide any guaranteed order. But I think that's fine - people know Maps (in general, unless e.g. they are specific ordered maps implemented as trees) aren't ordered and that's fine. It's like iteration order of a HashMap in Java - there's no guaranteed order but that doesn't make it not useful. We _could_ if we wanted impose an ordering here (e.g. by sorting they keys), but I doubt that's what most users would want. Also it would change the time complexity from O(N() to O(nlogn) (best case).

> On writing a table function to have the cross join output, note that the result of the function will be one column while the result of the cross join for two explodes will have two columns.

It wouldn't be too hard to extend table functions in the future to support output of multiple columns. I haven't seen much demand for this, but it's certainly doable. This functionality might also be useful, e.g. for exploding structs into multiple rows and other things (see 'inline" and 'stack' from Hive https://cwiki.apache.org/confluence/display/Hive/LanguageManual+UDF#LanguageManualUDF-Built-inTable-GeneratingFunctions(UDTF) ), but I don't recommend we do this yet.

I think a default behaviour of cross join semantics on the SELECT would also be bad from a complexity point of view (both algorithmic and user friendliness definitions of complexity ;) ).

In terms of algorithmic complexity: Cross joining has O(N*N) both time and space complexity. If we most users really want a zipjoin and they are having to do a cross join just in order to filter out different indexes to get a zipjoin then this could be a real problem as arrays get large.

In terms of user friendliness. It's not possible to have non column expressions in a WHERE clause so, aiui, the filtering would have to be done in a separate query, so you'd have something like

Query1:

CREATE STREAM S1 as SELECT INDEX_EXPLODE(a1) V!, INDEX_EXPLODE(a2) V2 FROM S0;
CREATE STREAM S2 as SELECT * FROM S1 WHERE V1->index == V2->index

Which I think also creates news topics for intermediate results? (Sorry, I am still new with ksql!)

big-andy-coates commented on design-proposals/klip-9-table-functions.md:
```diff
@@ -0,0 +1,206 @@
+extra FlatMap stage required by the query.
+* Amend the columns used in the final projection to reflect any generated columns created by a 
+FlatMap stage.
+* Implement the actual FlatMap logic. For each row, any table functions on that row will be evaluated.
```
> The number of rows returned from the FlatMap will be equal to the greatest of the number of rows returned from any of the table functions for that row.

As mentioned above, I think it should be the cartesian product.

purplefox replied:
I discussed this quite a bit with @derekjn and others and we could think of many more useful use cases where a zipjoin was desired, (e.g. the sensors example in the KLIP) but we couldn't come up with any for a cartesian join. If you can provide an example that would help me understand the utility of the suggestion a bit better :)

big-andy-coates commented on design-proposals/klip-9-table-functions.md:
```diff
@@ -0,0 +1,206 @@
+
+* `EXPLODE(ARRAY<V>)` -> Results in `N` rows (where `N` is number of elements in array) each with a column
+of type `V` holding the value of the element
+* `EXPLODE(MAP<K, V>)` -> Results in `N` rows (where `N` is number of elements in map) each with a column
```
FYI, I think we can augment this functionality with support for aliasing to flatten structured data into multiple columns, see. https://github.com/confluentinc/ksql/issues/3532.

big-andy-coates commented on design-proposals/klip-9-table-functions.md:
```diff
@@ -0,0 +1,206 @@
+
+### Implement framework of user defined table functions
+
+We will introduce annotations `@UdtfFactory` and `@UdtfDescription` in a similar way to UDAFs.
```
As per @agavra comment - I don't think we need a factory.  We should go more inline with UDFs.

And example UDTF here would help people to get a grip on what is being proposed.

purplefox replied:
I published a working prototype last well, with an example UDTF and working QTT test:

https://github.com/purplefox/ksql/pull/6

big-andy-coates commented on design-proposals/klip-9-table-functions.md:
```diff
@@ -0,0 +1,206 @@
+
+## Documentation Updates
+
+The following changes to the documentation will be made:
```
This docs section needs expanding as the requirements/approach solidify.  Ideally, this docs section should _be_ the text that will ultimately end up in the real docs.  Adding the documentation here allows those reviewing the KLIP to see the same docs the user will ultimately see.  If we can't understand if from the docs then either the docs ain't right, or the solution doesn't hang together.

purplefox commented on design-proposals/klip-9-table-functions.md:
````diff
@@ -0,0 +1,206 @@
+
+```
+CREATE STREAM exploded_stream AS
+    SELECT sensor_id, EXPLODE(readings) AS reading FROM batched_readings;
````
Supporting UDTFs in FROM clauses was my original suggestion and was discussed in the original thread here https://github.com/confluentinc/ksql/issues/527 - however, the consensus seemed to move on from there to UDTFs on the SELECT clause being a more pragmatic choice. 
I think it's fine to revisit this point, but I would like us to get some consensus on this as its such a fundamental point, and the prototype was based on the assumption that UDTFs on SELECTs is the way to go :)

purplefox replied:
>> `SELECT x.*, f.*, i.* FROM x LATERAL VIEW finiteUdft() AS f LATERAL VIEW infinitieUdtf() As i;`

Presumably this would result in a cross join? I doubt that's what most users would want.

I made this point in an earlier comment, but I think zip semantics are going to be more in demand, as yet I haven't heard a single concrete use case where a cartesian join wold be useful. Here's the example of zip I used earlier:

```
{"country":"usa",customer_names:["tim", "derek", "apurva"], customer_ages[99, 98, 97]}
```

expanding to:

```
{"country":"usa", customer_name:"tim", customer_age:99}
{"country":"usa", customer_name:"derek", customer_age:98}
{"country":"usa", customer_name:"apurva", customer_age:97}
```

How would you express a zip using udtfs on the FROM clause?

I don't see how it could be done without either a) creating a new JOIN type (ZIP) b) Implementing correlated subqueries (unintuitive, hard to grok) c) doing a cartesian join and a separate filtering step in a separate query (inefficient O(N*N) plus poor latency due to intermediate topic)

big-andy-coates replied:
Having a SELECT that returns multiple rows per input row feels wrong. So, it's pretty interesting to see this is the route that Postgres has gone! 

If we're going this route than I agree that a zip join seems more intuitive in the select expression.  I guess we can always support cartesian product joins via other syntax later if we need.

It seems the main fly in the ointment now is the unordered nature of MAPs. I definitely feels wrong to have the result of two exploded MAPs be dependent on the iteration order of Java HashMap, which might mean it changes depending on Java version.

big-andy-coates replied:
> Perhaps we should disallow EXPLODE taking a MAP directly

Yeah, I came to the same conclusion when thinking about them later.  

> Then we could have another tableFunction SORTED_PAIRS(MAP) which takes a map and outputs an array of pairs sorted by key.

I think this can be a generic UDF that converts map -> array.

> If you have just one exploded MAP in a SELECT you probably don't care about order of the pairs so UNSORTED_PAIRS allows you to this without the O(NlogN) penalty of sorting.

On exploded map, _with no other UDTFs_ is fine, but as soon as you mix in another UDTF, regardless of param type, then the ordering becomes an issue.

purplefox commented on design-proposals/klip-9-table-functions.md:
```diff
@@ -0,0 +1,206 @@
+or tables in the FROM clause. We do not think there is much value in doing this as the main driver
+for this KLIP is in supporting `explode` functionality which is more simply implemented on the
+SELECT clause.
+* Table functions that output more than one column. Some databases support the flattening of an array
```
My original suggestion was indeed to explode struct fields out into multiple columns, but I don't think that is necessarily needed as an initial step.

big-andy-coates replied:
Agreed, I think we can implement that flattening into multiple columns in a more generic way, i.e. not just for UDTFs but for flattening any complex type.

hjafarpour commented:
Hi @purplefox 
My question was in the context of having two explode calls with map input. I think the following example can provide more context. Let's consider we have two maps in a row:
```
 map1                              map2
--------                        ----------
k11 | v11                       k21 | v21 
k12 | v12                       k22 | v22
````
And we have two explodes calls in one select query:

```
SELECT id, EXPLODE(map1) AS c1, EXPLODE(map2) AS c2 FROM foo;
```
If we have cross join as default behavior the output rows will be as follows (note that the order may not be the same all the time as you pointed out above):

```
id        |      c1             |  c2          
----------------------------------------
1         |    <k11, v11>       |    <k21, v21>
1         |    <k11, v11>       |    <k22, v22>
1         |    <k12, v12>       |    <k21, v21>
1         |    <k12, v12>       |    <k22, v22>
```
The above results set will always contain the same rows (not considering the order of course) and hens is deterministic.

However, if we use the zipping behavior we can have any of the following four possible different results which makes it non deterministic:

```
id        |      c1             |  c2          
----------------------------------------
1         |    <k11, v11>       |    <k21, v21>
1         |    <k12, v12>       |    <k22, v22>
```
or
```
id        |      c1             |  c2            
----------------------------------------
1         |    <k11, v11>       |    <k22, v22>
1         |    <k12, v12>       |    <k21, v21>
```
or
```
id        |      c1             |  c2          
----------------------------------------
1         |    <k12, v12>       |    <k21, v21>
1         |    <k11, v11>       |    <k22, v22>
```
or
```
id        |      c1             |  c2         
----------------------------------------
1         |    <k12, v12>       |    <k22, v22>
1         |    <k11, v11>       |    <k21, v21>
```

purplefox commented:
> 
> The above results set will always contain the same rows (not considering the order of course) and hens is deterministic.

The rows created by a cross join may be deterministic but I don't think it means the results are particular useful (Well.. I'm sure there is a use case somewhere for this, but it doesn't seem obvious to me).

If a user is exploding multiple maps in a SELECT than I think it's more likely they want to zip based on a sorted order, in which case we could provide functions to convert the Map to a List of <K,V> and another to sort it.

purplefox commented:
> Also, I think this document would benefit if the details of the APIs we're to expose are fleshed out. I find a good way of doing this is to knock out a quick prototype once the discussion of how things should work die down. You can then use lessons learned from the prototype to document the APIs in here.

Andy, I think you may have missed the prototype I published last week for the team to look at, and the previous discussion.

big-andy-coates commented:
Hey @purplefox, nope I wasn't aware of the other thread.  I'm now all caught up ;)

It still feels weird / wrong to me to have UDTFs in the SELECT, but I can see that Postgres has paved the way with this, and Postgres also implements this as a ZIP join.  I'm fine with that as long as we're happy we can easily extend the functionality later if/when we decide to have cross joins.  

Multi-way UDTFs on MAPs still looks to be problematic, as @hjafarpour has highlighted with an example.  Non-deterministic output is problematic and unlikely to be of any use to users.  If users combine an EXPLODE(someMap) with any other UDTF then the output rows would be non-deterministic, potentially changing depending on JVM author or version. One solution to this would be to offer a variant of MAP UDTFs that takes an additional parameter that defines the order in which keys should be output, but such a solution would require _each_ UDTF to provide this overload. A more generic solution would be to provide a built-in UDF that converts a MAP into an ordered ARRAY of key-value pairs.  We might want to consider such a UDF part of the work for UDTFs to ensure we offer a rounded first drop of functionality.

That leaves the question of the API. Yes, at the moment an infinite UDTF would hold a StreamThread. But I don't think we should use that as a reason to not think about designing an API that could handle infinite UDTFs. Yes, it wouldn't be very efficient, but that doesn't mean it wouldn't be useful and we can always optimize later.  Hence I'd like to see more details on the UDTF API in this document.

It's really great to see this progressing - this is a much requested feature!

purplefox commented:
>> A more generic solution would be to provide a built-in UDF that converts a MAP into an ordered ARRAY of key-value pairs. We might want to consider such a UDF part of the work for UDTFs to ensure we offer a rounded first drop of functionality.

I think we're almost on the same page, as I commented and suggested the same thing at the same time you made this comment :)

purplefox commented:
> That leaves the question of the API. Yes, at the moment an infinite UDTF would hold a StreamThread. But I don't think we should use that as a reason to not think about designing an API that could handle infinite UDTFs. Yes, it wouldn't be very efficient, but that doesn't mean it wouldn't be useful and we can always optimize later. Hence I'd like to see more details on the UDTF API in this document.

This is a tricky one. If we wanted to create an API that works scalably and efficiently for <some hypothetical future underlying stream processing engine> then the form of that API would be radically different depending on whether the underlying stream engine uses a polling (pull) approach (like Rust futures/reactor) or a push approach (Vert.x/reactive streams/jdk 9 flow/spring reactor/etc).

Imho it's more likely that the <some hypothetical future underlying stream processing engine> would use a push approach as that's more common in the Java world. So if we decide on that we then have to decide on the form of the API - do we use JDK CompletableFuture or maybe ReactiveX observables http://reactivex.io/RxJava/javadoc/io/reactivex/Observable.html Reactive Streams https://www.reactive-streams.org JDK 9 flow https://docs.oracle.com/javase/9/docs/api/java/util/concurrent/Flow.html or something custom ?

We're likely to need back pressure so Reactive Streams seems a good choice. (Disclaimer: I was the Red Hat representative on the working group that wrote the Reactive Streams spec, this is now in JDK9 as flow). But reactive streams requires a fair bit of work on the function authors side to implement the correct APIs.

It should also be noted that any of these solutions also mean that the UDTF method can't be simple stateless function (like a UDF), as the future or subscriber or whatever is returned from the UDTF method will need to be called multiple times by the underlying framework.

To cut a long story short, the point I'm trying to make is I don't think there's a lot of point at this stage trying to write scalable streaming UDTF api that's going to work with this future streaming system. Whatever we choose now probably will need changing or adapting when that future system arrives (if it ever does). Also most UDTFs just need to return a List of results synchronously.

I think it would be more sensible to go with a simple synchronous API for now. And if the underlying streaming engine does support an async streaming approach in the future, it would be a simple matter to provide an adaptor that can be used automatically with any synchronous existing UDTF and convert it to the streaming API. At the point we could provide a new annotation @StreamingUDTF which could be used by users if they had need to create a UDTF with slow or infinite results.

purplefox commented:
I've made some updates as per review comments, to summarise:
* On the main point of contention, I kept the KLIP concentrating on table functions in SELECTs only
* I've added docs which should closely resemble the real docs
* I've added some more info on behaviour of multiple table functions in a SELECT
* A few other bits.

purplefox commented on design-proposals/klip-9-table-functions.md:
```diff
@@ -0,0 +1,347 @@
+Note that any scalar values in the select will be duplicated for each row in the output.
+
+We will also support multiple table functions in a SELECT clause. In this situation, the results of the
+table functions are `zipped` together.
```
There's an example already in the docs section :)

big-andy-coates replied:
Quick cut and paste then, for people reading this in a linear fashion?

purplefox commented on design-proposals/klip-9-table-functions.md:
```diff
@@ -0,0 +1,347 @@
+* `EXPLODE(ARRAY<V>)` -> Results in `N` rows (where `N` is number of elements in array) each with a column
+of type `V` holding the value of the element
+* `SORTED_PAIRS(MAP<K, V>)` -> Results in `N` rows (where `N` is number of entries in map), sorted by `K`
+* `UNSORTED_PAIRS(MAP<K, V>)` -> Results in `N` rows (where `N` is number of entries in map),
```
If you are exploding just one map you most likely don't care about ordering, and we don't want to impose an NlogN sorting overhead on those who don't need it.

purplefox commented on design-proposals/klip-9-table-functions.md:
```diff
@@ -0,0 +1,347 @@
+>|                        |                           |            | elements of the array. The output values are of the type of the     |                                     |
+>|                        |                           |            | array elements.                                                     |
+>+------------------------+---------------------------+------------+---------------------------------------------------------------------+
+>| SERIES                 | ``SERIES(col1, col2)`     | INT, INT   | This function takes two ints - start and end and output a range of  |
```
The description specifically says that the end bound is exclusive.

purplefox replied:
+1 on the BIGINT

purplefox commented on design-proposals/klip-9-table-functions.md:
```diff
@@ -0,0 +1,347 @@
+>Table functions can be thought of as analagous to the `FlatMap` operation commonly found in
+>functional programming or stream processing frameworks such as Kafka Streams.
+>
+>Table functions are used in the SELECT clause of a query. They cause the query to potentially
```
I will potentially change this to be less confusing ;)

purplefox commented on design-proposals/klip-9-table-functions.md:
```diff
@@ -0,0 +1,347 @@
+
+### Example of Table Function
+
+Add a new KSQL example for table functions
```
This is referring to the examples here https://docs.confluent.io/current/ksql/docs/tutorials/examples.html

big-andy-coates replied:
Maybe add the link to the doc?

purplefox commented on design-proposals/klip-9-table-functions.md:
```diff
@@ -0,0 +1,347 @@
+
+* `EXPLODE(ARRAY<V>)` -> Results in `N` rows (where `N` is number of elements in array) each with a column
+of type `V` holding the value of the element
+* `SORTED_PAIRS(MAP<K, V>)` -> Results in `N` rows (where `N` is number of entries in map), sorted by `K`
```
This is an interesting point. If we take this route then we only ever need an EXPLODE table function, all the other things that could otherwise be table functions could be implemented as normal UDFs which return a List<T>. E.g. splitting a string into words, or series().

We could go either route, but if we go the route of only using table functions to explode (that's the only thing a udf can't do), then we could disallow users from creating their own table functions (because they don't need, they can write a udf instead).

big-andy-coates replied:
I guess, but allowing custom UDTFs allows for users to make it easier / more succinct to write their SQL.  This is a good thing.

purplefox replied:
On the basis that SORTED_PAIRS can just be a UDF, the same logic also applies to SERIES and UNSORTED_PAIRS, so I'm going to remove these functions from the KLIP, and we'll only implement EXPLODE (we'll keep the UDTF functionality in though).

agavra commented on design-proposals/klip-9-table-functions.md:
```diff
@@ -0,0 +1,347 @@
+public class SplitString {
+
+    @Udtf(description="Splits a string into words")
+    public List<String> split(String input) {
```
probably goes without saying that we should make sure we support generics, structs and `@UdfSchemaProvider` in these APIs alongside `@UdtfParameter` or something like that  for the parameter documentation (and that all of these help contribute to the DESCRIBE FUNCTION call)

hpgrahsl commented:
...
> It would be good to think through that, because I imagine we will find valid use cases for cross joins.


...
> As yet, I haven't heard a convincing use case that would need cross join semantics, however if there is demand then it should be fairly easily supported by writing a new table function which does the cross product of the two arrays.

One simple example that comes to my mind would be that of representing a row containing product information with variants where you have e.g. `SIZE [S,M,L]` and `COLORS [R,G,B]` for T-shirts and then you would essentially want it to cross join to produce all possible variants. I'm sure you find several other examples as well, if you want to :)
